include/jemalloc/internal/cache_bin.h

```diff
@@ -427,6 +428,34 @@ cache_bin_full(cache_bin_t *bin) {
 	return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
 }
 
+/*
+ * Scans the allocated area of the cache_bin for the given pointer up to limit.
+ * Fires safety_check_fail if the ptr is found and returns true.
+ */
+JEMALLOC_ALWAYS_INLINE bool
+cache_bin_dalloc_safety_checks(cache_bin_t *bin, void *ptr) {
+        if (!config_debug || opt_debug_double_free_max_scan == 0) {
+                return false;
+        }
+
+        void **cur = bin->stack_head;
+        unsigned limit = opt_debug_double_free_max_scan;
+
+        while (limit != 0 && (uint16_t)(uintptr_t)cur != bin->low_bits_empty) {
```

Comment:
nit: Let's try optimizing a bit further -- for debug build it actually doesn't matter much; however we may add it to opt build later, also it can be a good example.
```
cache_bin_sz_t ncached = cache_bin_ncached_get_internal(bin, false);
unsigned max_scan = opt_debug_double_free_max_scan < ncached ? opt_debug_double_free_max_scan : ncached;

void **cur = bin->stack_head;
for (void **limit = cur + max_scan; cur < limit; cur++) {
    if (*cur == ptr) {
        ...
    }
}
```
This way we can avoid 1 branch check per iteration.

For the comments we always do `/* */`. It can be either one line, or multiple line format like you have for this function.

Reply:
addressed in [60a3574](https://github.com/jemalloc/jemalloc/pull/2310/commits/60a357485181d0cb1f45c6c04a7641992e9178ff)